internal/alb/sg/association.go

```diff
@@ -0,0 +1,266 @@
+type Association struct {
+	// We identify Association by LbID
+	LbID string
+
+	LbArn          string
+	LbPorts        []int64
+	LbInboundCIDRs types.Cidrs
+
+	// ExternalSGIDs are custom securityGroups intended to be attached to LoadBalancer.
+	// If customers specified these securityGroups via annotation on ingress, the ingress controller will then stop creating securityGroups for loadbalancer or ec2-instances.
+	ExternalSGIDs []string
+
+	Targets tg.TargetGroups
+}
+
+// AssociationController provides functionality to manage Association
+type AssociationController interface {
+	// Reconcile ensured the securityGroups in AWS matches the state specified by assocation.
+	Reconcile(*Association) error
+	
+	// Delete ensures the securityGroups created by ingress controller for specified LbID doesn't exists.
+	Delete(*Association) error
+}
+
+// NewAssociationController constructs a new association controller
+func NewAssociationController(store store.Storer, ec2 *albec2.EC2, elbv2 albelbv2.ELBV2API) AssociationController {
+	logger := log.New("sg.association")
+	lbAttachmentController := &lbAttachmentController{
+		elbv2:  elbv2,
+		ec2:    ec2,
+		logger: logger,
+	}
+	instanceAttachmentController := &instanceAttachmentController{
+		store:  store,
+		ec2:    ec2,
+		logger: logger,
+	}
+	sgController := &securityGroupController{
+		ec2:    ec2,
+		logger: logger,
+	}
+	namer := &namer{}
+	return &associationController{
+		lbAttachmentController:       lbAttachmentController,
+		instanceAttachmentController: instanceAttachmentController,
+		sgController:                 sgController,
+		namer:                        namer,
+		ec2:                          ec2,
+		logger:                       logger,
+	}
+}
+
+type associationController struct {
+	lbAttachmentController       LbAttachmentController
+	instanceAttachmentController InstanceAttachementController
+	sgController                 SecurityGroupController
+	namer                        Namer
+	ec2                          *albec2.EC2
+	logger                       *log.Logger
+}
+
+func (controller *associationController) Reconcile(association *Association) error {
+	if len(association.ExternalSGIDs) != 0 {
+		return controller.reconcileWithExternalSGs(association)
+	}
+	return controller.reconcileWithManagedSGs(association)
+}
+
+func (controller *associationController) Delete(association *Association) error {
+	return controller.deletedManagedSGs(association)
+}
+
+func (controller *associationController) reconcileWithExternalSGs(association *Association) error {
+	lbSGAttachment := &LbAttachment{
+		GroupIDs: association.ExternalSGIDs,
+		LbArn:    association.LbArn,
+	}
+	err := controller.lbAttachmentController.Reconcile(lbSGAttachment)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile external LoadBalancer securityGroup due to %s", err.Error())
+	}
+
+	err = controller.deletedManagedSGs(association)
```

Comment:
I think we need another one of these or just something to delete associations, we do not want to delete the externally manasged SG

Reply:
actually maybe this isnt needed at all for an externally managed sg, the only responsibility of the controller is to manage the SG attachmetns to the LB and instances right?